### ISSUE TITLE
Types of tenants proposal

### ISSUE THREAD

**Issue by MiltonLn** [django-tenants/django-tenants#30]
Hi guys.
In a SaaS architecture can be desirable to have a kind of tenant's types or clasificación.

**Comment by MiltonLn** [comment]
I'm sorry i have posted it by accident, so i will continue here my story...
Think about this:
You have an application with a few of modules, but you want to create some different types of tenants for tour customers with different number of modules, like a package of services.

So i was wondering if i can do a customization to django_tenants to support the types of tenants (if this isn't already created)
But i want to discuss this here for catch some ideas.
Initially i think i need to create various installed apps tuples on settings to support the types of tenants modules like
TYPEA_INSTALLED_APPS = (...)
TYPEB_INSTALLED_APPS = (...) etc..
And customize the migrate_schemas command ti support some like python migrate_schemas --tenants --type.
Also i think i need to customize the save of the tenant model for support the differentiation of tenant.

What do you think about this idea?
Thx

**Comment by tomturner** [comment]
That's sound like a good idea to me.

**Comment by ivome** [comment]
This could get quite tricky. Where do you plan to store the type? In the tenant model? What if a tenant changes its type? Then you would have to run a migration on each save of the model if the type changes. 
Some migrations in different APP sets might not be compatible. Or dependencies between models in apps that are part of different APP sets might cause problems. 
Something for handling different URLs also would have to be implemented. 

I am not sure if I would implement something like that on the installed apps layer. A custom app that defines a feature set for a tenant could be combined with a middleware that checks if the feature is enabled for the current tenant and then trigger some kind of action or raise an exception. This might be way easier and also more customizable depending on your business requirements. Maybe this could be implemented in a separate package? Not sure if this has to be part of this library, since it might get very specific?

**Comment by kosz85** [comment]
You have your points and we should take it into consideration. But about migrations, changes and messing with apps. You have same dangers now with moving apps between shared and tenant app settings. Developer has to decide himself, and plan application. I agree that it's quite advance, and shouldn't be default option.
You said that this includes improvements to other places (urls, middlewares, etc), so I think this lib should easily allow such toolset to exist, but I agree it should be separate project or sub-project that is plug-able. We should keep core of django-tenants simple. We may implement extensions interface for easier extension development. Then we could made different approaches as  extensions (domain-url middleware, url_path_prefix middleware, etc).

I would say that we should do such things in this repo, as separate repos would eventually detach from this one, but as kind of extension or set of additional libs (for use with celery, and things like this).

@tomturner What you think about it? Do we do such extensions as part of this repo in `extensions` dir or as separate projects? Or maybe you want to grow this project bigger and more powerful then django ;)

**Comment by tomturner** [comment]
I don't have the need for this in my sass project at the moment however I do understand that other people will.
So I think we should do this feature in this project so I welcome any pull requests
Before people work on this ticket it would be good if the design was posted on here first

**Comment by MiltonLn** [comment]
@tomturner please notice me for join the developers team for this feature.

I agree on discuss first the idea here, so, how can we start?

**Comment by kosz85** [comment]
I think it's wise to start from smth smaller first. You could plan what will be needed here, and then do for example #28.
In that way you can build more elastic environment of extensions, focusing on that issue and having in mind things you will need to do this one. Also you will see future conflicts between extensions.

**Comment by lorinkoz** [comment]
Building on @MiltonLn idea, some years afterwards, I do see the need for statically configured tenants. We sometimes need to have not only a shared tenant that can act as enterprise tenant, but also a finite (small) number of enterprise sites that respond to a custom behavior. I also believe that it could be better if we don't require the `public` schema to exist as an instance of `TENANT_MODEL`.

My proposal is to introduce static schemas configured in settings. They would take precedence over database schemas, and could help unifying the way in which we configure our tenants.

Please see the example:

```
TENANT_CONFIG = {
    'public': {
        'APPS': [
           ...
            'django_tenants',
            'django.contrib.admin',
            'django.contrib.auth',
            'django.contrib.contenttypes',
            'django.contrib.staticfiles',
            'common_app1',
            'common_app2',
            ...
        ],
    },
    'www': {
        'DOMAINS': ['my-site.com', 'www.my-site.com'],
        'URLCONF': 'enterprise.urls',
        'USES': ['public'],
        'APPS': [
            'django.contrib.sessions',
            'enterprise_app1',
            'enterprise_app2',
        ],
    },
    'help': {
        'DOMAINS': ['help.my-domain.com'],
        'URLCONF': 'help.urls',
        'USES': ['public'],
        'APPS': [
            'help_app1',
        ],
    },
    'default' : {
        'URLCONF': 'tenants.urls',
        'USES': ['public'],
        'APPS': [
            'django.contrib.sessions',
            'tenant_app1',
            'tenant_app2',
        ],
    }
}
```

`public` and `default` would be required in replacement to the current `SHARED_APPS` and `TENANT_APPS`. Note the dependency tree formed by `USES` that basically translates to using several schemas in the search path. We can have custom urlconfs for custom tenants like `www` and `help`. Notice in this example that the typical shared tenant is NOT acting as main tenant, and has truly become a shared tenant only.

Please let me summon @tomturner and @kosz85 for further discussion?
If this idea is acceptable and we can refine it to a feasible point, I'd be more than happy to work in it.

**Comment by tomturner** [comment]
Fixed Got there in the end #494
